// pouchdb-adapter-indexeddb/src/index.js
'use strict';
import { changesHandler } from 'pouchdb-utils';

import setup from './setup';

// API implementations
import info from './info';
import get from './get';
import getLocal from './getLocal';
import getAttachment from './getAttachment';
import bulkDocs from './bulkDocs';
import allDocs from './allDocs';
import changes from './changes';
import getRevisionTree from './getRevisionTree';
import doCompaction from './doCompaction';
import destroy from './destroy';
import {query, viewCleanup} from './find';
import purge from './purge';

import { DOC_STORE, META_LOCAL_STORE } from './util';

const ADAPTER_NAME = 'indexeddb';

// TODO: Constructor should be capitalised
const idbChanges = new changesHandler();

// A shared list of database handles
const openDatabases = {};

function IndexeddbPouch(dbOpts, callback) {

  if (dbOpts.view_adapter) {
    console.log('Please note that the indexeddb adapter manages _find indexes itself, therefore it is not using your specified view_adapter');
  }

  const api = this;

  // When the database is opened, we cache the "static" metadata fields i.e.
  // those that do not change when docs are updated. These are needed for some
  // async operations (e.g. preprocessing attachments) that cannot be performed
  // in the middle of a transaction due to the nature of the IndexedDB API, so
  // they need to know these settings without having to read from the object
  // store. "Dynamic" values like `doc_count` and `seq` must always be read
  // from the object store so that they are not stale.
  const CACHED_INFO_KEYS = ['db_uuid', 'idb_attachment_format'];
  const connectionInfo = {};

  function updateConnectionInfo(metadata) {
    for (const key of CACHED_INFO_KEYS) {
      connectionInfo[key] = metadata[key];
    }
  }

  // Wrapper that gives you an active DB handle. You probably want $t.
  const $ = function (fun) {
    return function () {
      const args = Array.prototype.slice.call(arguments);
      setup(openDatabases, api, dbOpts).then(function (res) {
        updateConnectionInfo(res.metadata);
        args.unshift(res.idb);
        fun.apply(api, args);
      }).catch(function (err) {
        const last = args.pop();
        if (typeof last === 'function') {
          last(err);
        } else {
          console.error(err);
        }
      });
    };
  };
  // the promise version of $
  const $p = function (fun) {
    return function () {
      const args = Array.prototype.slice.call(arguments);

      return setup(openDatabases, api, dbOpts).then(function (res) {
        updateConnectionInfo(res.metadata);
        args.unshift(res.idb);
        return fun.apply(api, args);
      });
    };
  };
  // Wrapper that gives you a safe transaction handle. It's important to use
  // this instead of opening your own transaction from a db handle got from $,
  // because in the time between getting the db handle and opening the
  // transaction it may have been invalidated by index changes.
  const $t = function (fun, stores, mode) {
    mode = mode || 'readonly';

    return function () {
      const args = Array.prototype.slice.call(arguments);
      const txn = {};
      setup(openDatabases, api, dbOpts).then(function (res) {
        updateConnectionInfo(res.metadata);
        txn.txn = res.idb.transaction(stores, mode);
      }).catch(function (err) {
        console.error('Failed to establish transaction safely');
        console.error(err);
        txn.error = err;
      }).then(function () {
        args.unshift(txn);
        fun.apply(api, args);
      });
    };
  };

  api._openTransactionSafely = function (stores, mode, callback) {
    $t(function (txn, callback) {
      callback(txn.error, txn.txn);
    }, stores, mode)(callback);
  };

  api._getMetadata = $t(function (txn, cb) {
    const metaStore = txn.txn.objectStore(META_LOCAL_STORE);
    metaStore.get(META_LOCAL_STORE).onsuccess = function (e) {
      cb(null, e.target.result);
    };
  }, [META_LOCAL_STORE]);

  api._remote = false;
  api.type = function () { return ADAPTER_NAME; };

  api._id = function (cb) {
    api._getMetadata(function (err, metadata) {
      cb(null, metadata.db_uuid);
    });
  };

  api._info = function (cb) {
    api._getMetadata(function (err, metadata) {
      return info(metadata, cb);
    });
  };

  api._get = $t(get, [DOC_STORE]);
  api._getLocal = $t(function (txn, id, callback) {
    return getLocal(txn, id, api, callback);
  }, [META_LOCAL_STORE]);

  api._bulkDocs = $(function (_, req, opts, callback) {
    bulkDocs(api, req, opts, connectionInfo, dbOpts, idbChanges, callback);
  });

  api._allDocs = $t(function (txn, opts, cb) {
    const metaStore = txn.txn.objectStore(META_LOCAL_STORE);
    metaStore.get(META_LOCAL_STORE).onsuccess = function (e) {
      const metadata = e.target.result;
      allDocs(txn, metadata, opts, cb);
    };
  }, [DOC_STORE, META_LOCAL_STORE]);

  api._getAttachment = getAttachment;

  api._changes = $t(function (txn, opts) {
    changes(txn, idbChanges, api, dbOpts, opts);
  }, [DOC_STORE]);

  api._getRevisionTree = $t(getRevisionTree, [DOC_STORE]);
  api._doCompaction = $t(doCompaction, [DOC_STORE], 'readwrite');

  api._customFindAbstractMapper = {
    query: $p(query),
    viewCleanup: $p(viewCleanup)
  };

  api._destroy = function (opts, callback) {
    return destroy(dbOpts, openDatabases, idbChanges, callback);
  };

  api._close = $(function (db, cb) {
    delete openDatabases[dbOpts.name];
    db.close();
    cb();
  });

  // Closing and re-opening the DB re-generates native indexes
  api._freshen = function () {
    return new Promise(function (resolve) {
      api._close(function () {
        $(resolve)();
      });
    });
  };

  api._purge = $t(purge, [DOC_STORE], 'readwrite');

  // TODO: this setTimeout seems nasty, if its needed lets
  // figure out / explain why
  setTimeout(function () {
    callback(null, api);
  });
}

// TODO: this isnt really valid permanently, just being lazy to start
IndexeddbPouch.valid = function () {
  return true;
};

export default function (PouchDB) {
  PouchDB.adapter(ADAPTER_NAME, IndexeddbPouch, true);
}
